java/vector/src/main/codegen/templates/DenseUnionVector.java

```diff
@@ -0,0 +1,887 @@
+
+      int [] typeCounts = new int[MinorType.values().length];
+      int [] typeStarts = new int[Types.MinorType.values().length];
+      for (int i = 0; i < typeCounts.length; i++) {
+        typeCounts[i] = 0;
+        typeStarts[i] = -1;
+      }
+
+      for (int i = startIndex; i < startIndex + length; i++) {
+        int ord = typeBuffer.getByte(i);
+        to.offsetBuffer.setInt((i - startIndex) * OFFSET_WIDTH, typeCounts[ord]);
+        typeCounts[ord] += 1;
+        if (typeStarts[ord] == -1) {
+          typeStarts[ord] = offsetBuffer.getInt(i * OFFSET_WIDTH);
+        }
+      }
+      to.setValueCount(length);
+
+      // transfer vector values
+      for (int i = 0; i < MinorType.values().length; i++) {
+        if (typeCounts[i] > 0 && typeStarts[i] != -1) {
+          internalTransferPairs[i].splitAndTransfer(typeStarts[i], typeCounts[i]);
+        }
+      }
+    }
+
+    @Override
+    public ValueVector getTo() {
+      return to;
+    }
+
+    @Override
+    public void copyValueSafe(int from, int to) {
+      this.to.copyFrom(from, to, DenseUnionVector.this);
+    }
+  }
+
+  @Override
+  public FieldReader getReader() {
+    if (reader == null) {
+      reader = new DenseUnionReader(this);
+    }
+    return reader;
+  }
+
+  public FieldWriter getWriter() {
+    if (writer == null) {
+      writer = new DenseUnionWriter(this);
+    }
+    return writer;
+  }
+
+  @Override
+  public int getBufferSize() {
+    if (valueCount == 0) { return 0; }
+
+    return valueCount * TYPE_WIDTH + valueCount * OFFSET_WIDTH + internalStruct.getBufferSize();
+  }
+
+  @Override
+  public int getBufferSizeFor(final int valueCount) {
+    if (valueCount == 0) {
+      return 0;
+    }
+
+    long bufferSize = 0;
+    for (final ValueVector v : (Iterable<ValueVector>) this) {
+      bufferSize += v.getBufferSizeFor(valueCount);
+    }
+
+    return (int) bufferSize + valueCount * TYPE_WIDTH + valueCount * OFFSET_WIDTH;
+  }
+
+  @Override
+  public ArrowBuf[] getBuffers(boolean clear) {
+    List<ArrowBuf> list = new java.util.ArrayList<>();
+    setReaderAndWriterIndex();
+    if (getBufferSize() != 0) {
+      list.add(typeBuffer);
+      list.add(offsetBuffer);
+      list.addAll(java.util.Arrays.asList(internalStruct.getBuffers(clear)));
+    }
+    if (clear) {
+      valueCount = 0;
+      typeBuffer.getReferenceManager().retain();
+      typeBuffer.close();
+      typeBuffer = allocator.getEmpty();
+      offsetBuffer.getReferenceManager().retain();
+      offsetBuffer.close();
+      offsetBuffer = allocator.getEmpty();
+    }
+    return list.toArray(new ArrowBuf[list.size()]);
+  }
+
+  @Override
+  public Iterator<ValueVector> iterator() {
+    List<ValueVector> vectors = org.apache.arrow.util.Collections2.toList(internalStruct.iterator());
+    return vectors.iterator();
+  }
+
+  private ValueVector getVector(int index) {
+    int type = typeBuffer.getByte(index * TYPE_WIDTH);
+    if (type < Types.MinorType.values().length) {
```

Comment:
this still doesn't conform to the spec as written.  There can pontentially be more then one vector of each primitive type (e.g. 2 IntVectors).  Again we should try to close on this on the mailing list, but I thought last time it was discussed we agreed to keep it general.

Reply:
Sure. I will fix this.

I am a little curious about in what scenario do we need to have 2 IntVectors? Two IntVectors with different meta data?

Another question is, do we need to fix this problem for the (sparse) UnionVector in a follow up issue?

Reply:
I have revised to code to support more than one vector for each vector type. Please take a look.

Reply:
I will try to review over the weekend.  To answer the question on use-case Arrow's unions underlying vectors are named meaning they could have two integers with different semantics.  Another example of a data format that allows this is [protobufs with oneof fields](https://developers.google.com/protocol-buffers/docs/proto).  Whether this is a common case is up for debate (I can't think of a good use-case at the moment) and the last time it was discussed on the mailing list the most common case is "json" representations which would exclude the duplicate integer use-cases.  This is how the SparseUnionVector is implemented in Java.

Reply:
I see. Thanks for the clarification.Have a great weekend.